### PR TITLE
Internal: Throw exception when binding a non-existing prop [EDS-297]

### PR DESCRIPTION
**Pull request by StyleShit**

### CODE REVIEW

File: modules/atomic-widgets/base/atomic-control-base.php [hunk 1]
```diff
@@ -21,6 +21,10 @@ protected function __construct( string $prop_name ) {
 		$this->bind = $prop_name;
 	}
 
+	public function get_bind() {
```
Comment (Nevoss) [inline modules/atomic-widgets/base/atomic-control-base.php:24]: use in `jsonSerialize`

File: tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php [hunk 2]
```diff
@@ -79,4 +84,165 @@ public function test_get_props_schema__is_serializable() {
...
+
+		// Expect.
+		$this->expectException( \Exception::class );
+		$this->expectExceptionMessage( 'Prop `not-in-schema` is not defined in the schema of `test-widget`. Did you forget to define it?' );
+
+		// Act.
+		$widget->get_atomic_controls();
+	}
+
+	public function test_get_atomic_controls__throws_when_plain_object_control_is_not_in_schema() {
```
Comment (Nevoss) [inline tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php:124]: Remove

File: tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php [hunk 3]
```diff
@@ -79,4 +84,165 @@ public function test_get_props_schema__is_serializable() {
...
+
+		// Expect.
+		$this->expectException( \Exception::class );
+		$this->expectExceptionMessage( 'Control is missing a bound prop from the schema.' );
+
+		// Act.
+		$widget->get_atomic_controls();
+	}
+
+	public function test_get_atomic_controls__throws_when_plain_object_control_has_empty_bind() {
```
Comment (Nevoss) [inline tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php:163]: Remove

File: tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php [hunk 4]
```diff
@@ -79,4 +84,165 @@ public function test_get_props_schema__is_serializable() {
...
+		$controls_definitions = [
+			// Top-level control
+			Textarea_Control::bind_to( 'text' ),
+
+			// Control in section
+			Section::make()->set_items( [
+				Select_Control::bind_to( 'select' )
+			] ),
+
+			// Plain object control
```
Comment (Nevoss) [inline tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php:197]: Remove

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: modules/atomic-widgets/base/atomic-widget-base.php [hunk 5]
```diff
@@ -19,7 +20,42 @@ public function __construct( $data = [], $args = null ) {
...
+
+		$schema = static::get_props_schema();
+
+		foreach ( $controls as $control ) {
+			if ( $control instanceof Section ) {
+				$this->validate_controls( $control->get_items() );
+				continue;
+			}
+
+			$prop_name = $control instanceof Atomic_Control_Base
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply modules/atomic-widgets/base/atomic-widget-base.php:44]: Done

File: modules/atomic-widgets/base/atomic-widget-base.php [hunk 6]
```diff
@@ -19,7 +20,42 @@ public function __construct( $data = [], $args = null ) {
...
+				? $control->get_bind()
+				: $control['value']['bind'] ?? null; // TODO: Do we want to support plain objects?
```
[the remark this answers is not in the document]
Comment (StyleShit) [reply modules/atomic-widgets/base/atomic-widget-base.php:46]: Done

File: modules/atomic-widgets/base/atomic-control-base.php (hunk 1 above)
Comment (StyleShit) [reply modules/atomic-widgets/base/atomic-control-base.php:24]: Done

File: tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php (hunk 2 above)
Comment (StyleShit) [reply tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php:124]: Done

File: tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php (hunk 3 above)
Comment (StyleShit) [reply tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php:163]: Done

File: tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php (hunk 4 above)
Comment (StyleShit) [reply tests/phpunit/elementor/modules/atomic-widgets/test-atomic-widget-base.php:197]: Done

File: modules/atomic-widgets/base/atomic-widget-base.php [hunk 7]
```diff
@@ -19,7 +19,60 @@ public function __construct( $data = [], $args = null ) {
...
+		return $valid_controls;
+	}
+
+	private function safe_throw( string $message ) {
+		if ( ! defined( 'ELEMENTOR_DEBUG' ) || ! ELEMENTOR_DEBUG ) {
+			return;
+		}
+
+		throw new \Exception( $message );
+	}
```
Comment (Nevoss) [inline modules/atomic-widgets/base/atomic-widget-base.php:73]: Maybe create general util for other use cases and classes

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: modules/atomic-widgets/base/atomic-widget-base.php (hunk 7 above)
Comment (StyleShit) [reply modules/atomic-widgets/base/atomic-widget-base.php:73]: NO

**Review by Nevoss — APPROVED** [review]

**Pull request merged**